### ISSUE TITLE
generates the wrong code?

### ISSUE THREAD

**Issue by falafflepotatoe** [jashkenas/coffeescript#4168]
not 100% on this one but playing on the coffeescript site these two funcs generate the different code

output from cs.org: http://goo.gl/8bIq2S
code 1:

    remove=(fn)-> 
       for v,i in coll by -1
          if fn(v)
            coll.splice(i,1)

code 2 (doesnt work):

     remove=(fn)-> coll.splice(i,1) if fn(v) for v,i in coll by -1

then `remove((x)->x.done==true)`

full snippet:

    #setup, collection of todos {id: 1, done: true|false}
    coll=[]
    coll.push {id:i, done:!!(Math.random()*2|0)} for i in [1..9]
    
    #so these work (they are all the same code)
    remove  = (fn) -> for v,i in coll by -1 then if fn(v) then coll.splice(i,1)
    remove2= (fn) -> 
       for v,i in coll by -1
          if fn(v)
            coll.splice(i,1)
     #doesnt work
     remove3=(fn)-> coll.splice(i,1) if fn(v) for v,i in coll by -1
     
     remove3( (x)->x.done==true )

tl:dr i thought this was the right syntax, apparently it is not? ex: 

`alert v if v%3==0 for v in [1..10]`

**Comment by lydell** [comment]
```coffee
console.log n if n % 3 == 0 for n in [1..10]
# ... is equivalent to ...
console.log n if (n % 3 == 0 for n in [1..10])
# ... which basically means:
n = 11; console.log n

# This seems to be what you're looking for:
(console.log n if n % 3 == 0) for n in [1..10]
# ... which can be written like this:
console.log n for n in [1..10] when n % 3 == 0
```

**Comment by falafflepotatoe** [comment]
@vendethiel thanks!
